File: python_cpp/uPIMulator_backend/src/simulator/sram/lock.cc
#include "simulator/sram/lock.h"

namespace upmem_sim::simulator::sram {

void Lock::acquire(ThreadID id) {
  assert(can_acquire());
  id_ = new ThreadID(id);
}

void Lock::release(ThreadID id) {
  assert(can_release(id));

  delete id_;
  id_ = nullptr;
}

}  // namespace upmem_sim::simulator::sram

File: python_cpp/uPIMulator_frontend/src/iss/sram/lock.py
from typing import Optional


class Lock:
    def __init__(self):
        self._id: Optional[int] = None

    def can_acquire(self):
        return self._id is None

    def acquire(self, id_: int):
        assert self.can_acquire()
        self._id = id_

    def can_release(self, id_: int):
        return self._id is None or self._id == id_

    def release(self, id_: int):
        assert self.can_release(id_)
        self._id = None
